Fixes #29049 - hide deprecated images & extends projects in GCE

kgaikwad opened the pull request:
With this commit, it appends Google extra global projects list
with 'rhel-sap-cloud' project to show images as like Google
compute engine. Also, adds a method to override image list from
google compute by adding image family names to list.

+ adding @jyejare & @ares into the loop.

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

kgaikwad committed:
Fixes #29049 - hide deprecated images & extends projects in GCE

With this commit, it appends Google extra global projects list
with 'rhel-sap-cloud' project to show images as like Google
compute engine. Also, adds a method to override image list from
google compute by adding image family names to list.

kgaikwad commented:
Test failure is irrelevant to the changes

ares commented:
I assume this is OK for the first version. I don't have time to test but code looks good. Few suggestions (for future if it's too much work for this PR):

* we can easily support wildcards,e.g. `centos*`, you can reuse existing method that we have for `Setting` https://github.com/theforeman/foreman/blob/6020f357300b196eef7abcaae4a0d9f4945d0039/app/models/setting.rb#L209-L213
* we should expose the `image_families_to_filter` to the user, perhaps just by adding a textfield in GCE form (and API), users may want to adjust what they see without overriding the method from any plugin

ezr-ondrej commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -154,7 +154,18 @@ def create_volumes(args)
     def available_images
-      client.images
+      images_list = client.images.current
+      if image_families_to_filter.present?
```
can this be `nil`? if not:
```suggestion
      if image_families_to_filter.any?
```
sounds bit better

ezr-ondrej reviewed (dismissed):
Small nitpick, apart of it looks good, but I guess there is a followup for this? or am I able to test it somehow now? (As @ares is mentioning, there is no way to manage `image_families`.

kgaikwad committed:
Refs #29049 - reuse setting's regex method for gce image filtering

kgaikwad commented:
@ares 
> Few suggestions (for future if it's too much work for this PR):
> * we can easily support wildcards,e.g. `centos*`, you can reuse existing method that we have for `Setting` https://github.com/theforeman/foreman/blob/6020f357300b196eef7abcaae4a0d9f4945d0039/app/models/setting.rb#L209-L213

:+1:  Instead of using this method I have used other method  `convert_array_to_regexp` method.
https://github.com/theforeman/foreman/blob/6020f357300b196eef7abcaae4a0d9f4945d0039/app/models/setting.rb#L215 
Please correct me if this is not expected here.

> * we should expose the `image_families_to_filter` to the user, perhaps just by adding a textfield in GCE form (and API), users may want to adjust what they see without overriding the method from any plugin

Surely, as this PR change needs to be cherry-pick in 1.24, I added new ticket for this improvement.
https://projects.theforeman.org/issues/29182

kgaikwad commented:
> but I guess there is a followup for this? or am I able to test it somehow now? (As @ares is mentioning, there is no way to manage `image_families`.

@ezr-ondrej, with the current changes, you can only do testing around hide deprecated images from images list. I thought to override this image_families_to_filter method in theme plugin to show only RHEL images so you won't be able to test this now.

ezr-ondrej commented on app/models/compute_resources/foreman/model/gce.rb:
```diff
@@ -154,7 +154,19 @@ def create_volumes(args)
+    # override this method to filter list by adding image family names
+    # Example - ['rhel', 'centos']
+    def image_families_to_filter
+      []
```
Would it make sense to have some registration to this? Take it as nitpick, if this is not meant for major usage, no need for it now.
I.E like:
```ruby
  def self.image_families_to_filter
      @image_families_to_filter ||= []
  end
  def self.register_family_for_image_filter(family)
    image_families_to_filter << family
  end
```
```suggestion
      self.class.image_families_to_filter
```

kgaikwad replied:
Done!

ezr-ondrej reviewed (commented):
could we add a test for the `available_images` method?
To test the case we stub the `image_families` to return some content.
I would expect it to be a simple test just to ensure the filtering is working as expected.

kgaikwad committed:
Refs #29049 - add method to register image families & tests cases

kgaikwad commented:
> could we add a test for the `available_images` method?
> To test the case we stub the `image_families` to return some content.
> I would expect it to be a simple test just to ensure the filtering is working as expected.

@ezr-ondrej, @shiramax,
Updated PR with suggested changes and added two test cases around `available_images` method.

ezr-ondrej reviewed (approved):
LGTM! The test looks like testing it, looking forward to see it in action.

ezr-ondrej commented:
Thanks @kgaikwad :+1:

tbrisker commented:
source release failure seems related: https://community.theforeman.org/t/foreman-develop-source-release-534-failed/17810/2
